components/ads-client/src/error.rs

```diff
@@ -0,0 +1,237 @@
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum BuildRequestError {
+    #[error("Could not build request with empty placement configs")]
+    EmptyConfig,
+
+    #[error("Duplicate placement_id found: {placement_id}. Placement_ids must be unique.")]
+    DuplicatePlacementId { placement_id: String },
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum BuildPlacementsError {
+    #[error("Duplicate placement_id found: {placement_id}. Placement_ids must be unique.")]
+    DuplicatePlacementId { placement_id: String },
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum FetchAdsError {
+    #[error("URL parse error: {0}")]
+    UrlParse(#[from] url::ParseError),
+
+    #[error("Error sending request: {0}")]
+    Request(#[from] viaduct::Error),
+
+    #[error("JSON error: {0}")]
+    Json(#[from] serde_json::Error),
+
+    #[error("Could not fetch ads, MARS responded with: {0}")]
+    HTTPError(#[from] HTTPError),
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum EmitTelemetryError {
+    #[error("URL parse error: {0}")]
+    UrlParse(#[from] url::ParseError),
+
+    #[error("Error sending request: {0}")]
+    Request(#[from] viaduct::Error),
+
+    #[error("JSON error: {0}")]
+    Json(#[from] serde_json::Error),
+
+    #[error("Could not fetch ads, MARS responded with: {0}")]
+    HTTPError(#[from] HTTPError),
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum CallbackRequestError {
+    #[error("URL parse error: {0}")]
+    UrlParse(#[from] url::ParseError),
+
+    #[error("Error sending request: {0}")]
+    Request(#[from] viaduct::Error),
+
+    #[error("JSON error: {0}")]
+    Json(#[from] serde_json::Error),
+
+    #[error("Could not fetch ads, MARS responded with: {0}")]
+    HTTPError(#[from] HTTPError),
+
+    #[error("Callback URL missing: {message}")]
+    MissingCallback { message: String },
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum RecordImpressionError {
+    #[error("Callback request to MARS failed: {0}")]
+    CallbackRequest(#[from] CallbackRequestError),
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum RecordClickError {
+    #[error("Callback request to MARS failed: {0}")]
+    CallbackRequest(#[from] CallbackRequestError),
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum ReportAdError {
+    #[error("Callback request to MARS failed: {0}")]
+    CallbackRequest(#[from] CallbackRequestError),
+}
+
+#[derive(Debug, thiserror::Error)]
+pub enum HTTPError {
+    #[error("Validation error ({code}): {message}")]
+    Validation { code: u16, message: String },
+
+    #[error("Bad request ({code}): {message}")]
+    BadRequest { code: u16, message: String },
+
+    #[error("Server error ({code}): {message}")]
+    Server { code: u16, message: String },
+
+    #[error("Unexpected error ({code}): {message}")]
+    Unexpected { code: u16, message: String },
+}
+
+pub fn check_http_status_for_error(response: &Response) -> Result<(), HTTPError> {
+    let status = response.status;
+    if status >= 400 {
```

Comment:
this seems a little risky wrt other codes - eg, 3xx, even 1xx, if seen, probably shouldn't be treated as success? Is anything other than 200 really expected here?